### ISSUE TITLE
UnsatisfiedLinkError ,when i execute “Realm.getDefaultInstance()”

### ISSUE THREAD

**Issue by qsxiaogang** [realm/realm-java#2927]
crash log:

java.lang.UnsatisfiedLinkError: No implementation found for long io.realm.internal.w.a(java.lang.String, byte[]) (tried Java_io_realm_internal_w_a and Java_io_realm_internal_w_a__Ljava_lang_String_2_3B)
                                                     at io.realm.internal.w.a(Native Method)
                                                     at io.realm.internal.w.<init>(SharedGroup.java:63)
                                                     at io.realm.internal.z.<init>(SharedGroupManager.java:49)
                                                     at io.realm.d.<init>(BaseRealm.java:79)
                                                     at io.realm.n.<init>(Realm.java:138)
                                                     at io.realm.n.b(Realm.java:248)
                                                     at io.realm.n.a(Realm.java:228)
                                                     at io.realm.t.a(RealmCache.java:126)
                                                     at io.realm.n.v(Realm.java:174)
                                                     at com.ccclubs.dk.ui.cab.SelectAddressActivity.init(SelectAddressActivity.java:105)
                                                     at com.ccclubs.common.base.BaseActivity.onCreate(BaseActivity.java:100)
                                                     at android.app.Activity.performCreate(Activity.java:6071)
                                                     at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1123)
                                                     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2561)
                                                     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2677)
                                                     at android.app.ActivityThread.access$800(ActivityThread.java:178)
                                                     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1539)
                                                     at android.os.Handler.dispatchMessage(Handler.java:111)
                                                     at android.os.Looper.loop(Looper.java:194)
                                                     at android.app.ActivityThread.main(ActivityThread.java:5773)
                                                     at java.lang.reflect.Method.invoke(Native Method)
                                                     at java.lang.reflect.Method.invoke(Method.java:372)
                                                     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1005)
                                                     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:800)

```java

in my Application

    RealmConfiguration config = DatabaseHelper.getConfig(this);
    Realm.setDefaultConfiguration(config);

  public static RealmConfiguration getConfig(Context mContext) {
    RealmConfiguration config = new RealmConfiguration.Builder(mContext).name(dbName)
        .schemaVersion(dbVersion)
        .deleteRealmIfMigrationNeeded()
        .build();
    return config;
  }

the crash line:
// line 105
realm = Realm.getDefaultInstance();

```

Realm version(s): 1.0.0

Android Studio version: 2.1.1

Which Android version and device: 5.1 (API 22)

have any suggestion ?

tips: i have two versions in gradle,they have two different applicationId,when debug,it run as well,but after signed , crashed.

**Comment by nhachicha** [comment]
Hi @qsxiaogang 
can you share your Gradle & ProGuard configuration please?
Cheers

**Comment by qsxiaogang** [comment]
project Gradle:
```
buildscript {
  repositories {
    jcenter()
  }
  dependencies {
    classpath 'com.android.tools.build:gradle:2.0.0'
    classpath 'me.tatarka:gradle-retrolambda:3.2.5'
    classpath "io.realm:realm-gradle-plugin:1.0.0"
    // NOTE: Do not place your application dependencies here; they belong
    // in the individual module build.gradle files
  }
}

ext {
  configuration = [//            package          : "com.ccclubs.dk",
                   compileSdkVersion: 23,
                   buildToolsVersion: "23.0.2",
                   minSdk           : 13,
                   targetSdk        : 23,
                   version_code     : 1,
                   version_name     : "0.0.1"]

  libraries = [support    : "23.2.1",
               design     : "23.2.1",
               gson       : "2.4",
               analytics  : "5.6.1",
               //               okhttp2    : "2.1.0",
               okhttp3    : "3.2.0",
               picasso    : "2.5.2",
               retrofit   : "2.0.2",
               rxjava     : "1.1.3",
               rxandroid  : "1.1.0",
               rxlifecycle: "0.5.0",
               //            dagger            : "2.0",
               //            javax_annotation  : "10.0-b28",
               butterknife: "7.0.1",
               sweetalert : "1.3",
               rebound    : "0.3.8",
               greendao   : "2.2.0",
               eventbus   : "3.0.0"
               //glide             : "3.6.1",
               //            recycler_animators: "2.1.0",
               //circleimageview   : "2.0.0"
  ]
}

allprojects {
  repositories {
    jcenter()
    maven { url "https://jitpack.io" }
  }
}

task clean(type: Delete) {
  delete rootProject.buildDir
}
```
app Gradle:
```
apply plugin: 'com.android.application'
apply plugin: 'me.tatarka.retrolambda'
apply plugin: 'realm-android'
....

```

>A ProGuard configuration is provided as part of the Realm library. This means that you don’t need to add any Realm specific rules to your ProGuard configuration.

so there is nothing about Realm in my ProGuard.

**Comment by qsxiaogang** [comment]
```
productFlavors {
      penglong {
        applicationId 'com.ccclubs.dk'
        versionCode 246
        versionName "2.4.6"
        manifestPlaceholders =
            [application_id: "com.ccclubs.dk", UMENG_APPVALUE: "55d1a50b67e58e740sdfsa0f90", JPUSH_APPVALUE: "3aa095f7180ea39c6sdfc4", AMAP_APPVALUE: "94a32d8a2afe712126af2600bc595a", UMENG_CHANNEL: "default"]
        buildConfigField "boolean"
        buildConfigField "String", "BASE_URL", "\"http://app1.ccclubs.com/app/\""
      }

      plBeta {
        applicationId 'com.ccclubs.dk.better'
        versionCode 215
        versionName "2.1.5"
        // release : AMAP_APPVALUE:"0a9ad9fd4f35daf3e0e6077e9e692144" | debug: AMAP_APPVALUE:"1a2beaca10ee95fb96a5d13383ffb58e"
        manifestPlaceholders =
            [application_id: "com.ccclubs.dk.better", UMENG_APPVALUE: "561f09d667e58effasdf", JPUSH_APPVALUE: "88dd78af73dcb9dd27fasdf", AMAP_APPVALUE: "1a2beaca10ee95fb96a5d13383fsdf", UMENG_CHANNEL: "default"]
       
        buildConfigField "boolean", 
        buildConfigField "String", "BASE_URL", "\"http://pltest1.app.ccclubs.com/\"
      }
    }
  }
}

dependencies {
  compile fileTree(include: '*.jar', dir: 'libs')
  compile 'com.nineoldandroids:library:2.4.0'
  compile "com.google.code.gson:gson:${libs.gson}"
  compile "com.umeng.analytics:analytics:${libs.analytics}"
  compile "com.squareup.picasso:picasso:${libs.picasso}"
  compile "com.facebook.rebound:rebound:${libs.rebound}"
  testCompile 'junit:junit:4.12'
  compile project(':commonlibs')
//  compile 'com.github.qsxiaogang:CommonCoreLib:0.0.1'
}
```

**Comment by qsxiaogang** [comment]
```
productFlavors {
      peng {
        applicationId 'com.ccclubs.dk'
        versionCode 246
        versionName "2.4.6"
        manifestPlaceholders =
            [application_id: "com.ccclubs.dk", UMENG_APPVALUE: "55d1a50b267e58e740sdfsa0f90", JPUSH_APPVALUE: "3aa095f7180eaff39c6sdfc4", AMAP_APPVALUE: "94a32d8a2afe7121f26affa2600bc595a", UMENG_CHANNEL: "default"]
        buildConfigField "boolean"
        buildConfigField "String", "BASE_URL", "\"http://app1.ccclubs.com/app/\""
      }

      plBeta {
        applicationId 'com.ccclubs.dk.better'
        versionCode 215
        versionName "2.1.5"
        manifestPlaceholders =
            [application_id: "com.ccclubs.dk.better", UMENG_APPVALUE: "561ff09d667e58effasdf", JPUSH_APPVALUE: "88dd78af73dcbf9dd27fasdf", AMAP_APPVALUE: "1a2beaca10ee95fb96a5d13f383fsdf", UMENG_CHANNEL: "default"]

        buildConfigField "boolean", 
        buildConfigField "String", "BASE_URL", "\"http://pltest1.app.ccclubs.com/\"
      }
    }
  }
}

dependencies {
  compile fileTree(include: '*.jar', dir: 'libs')
  compile 'com.nineoldandroids:library:2.4.0'
  compile "com.google.code.gson:gson:${libs.gson}"
  compile "com.umeng.analytics:analytics:${libs.analytics}"
  compile "com.squareup.picasso:picasso:${libs.picasso}"
  compile "com.facebook.rebound:rebound:${libs.rebound}"
  testCompile 'junit:junit:4.12'
  compile project(':commonlibs')
//  compile 'com.github.qsxiaogang:CommonCoreLib:0.0.1'
}
```

**Comment by Zhuinden** [comment]
With these names

    at io.realm.internal.w.a(Native Method)
    at io.realm.internal.w.(SharedGroup.java:63)
    at io.realm.internal.z.(SharedGroupManager.java:49)

I'd think the issue is that Proguard ate your stuff. Maybe the Gradle-plugin's proguard config broke?
    
    #realm
    -keepnames public class * extends io.realm.RealmObject
    -keep @io.realm.annotations.RealmModule class *
    -keep class io.realm.** { *; }
    -dontwarn javax.**
    -dontwarn io.realm.**

    #realm 0.84.1+
    -keep class io.realm.annotations.RealmModule
    -keep @io.realm.annotations.RealmModule class *
    -keep class io.realm.internal.Keep
    -keep @io.realm.internal.Keep class *
    -dontwarn javax.**
    -dontwarn io.realm.**

    #realm 0.89.0+
    -keep class io.realm.RealmCollection
    -keep class io.realm.OrderedRealmCollection

I've been using this Proguard config and it seems to be working

**Comment by qsxiaogang** [comment]
@Zhuinden thanks for your suggestion,i will try soon.

**Comment by qsxiaogang** [comment]
@Zhuinden thank you.
i add this to my ProGurad, and it works all right.
```
#realm
-keepnames public class * extends io.realm.RealmObject
-keep @io.realm.annotations.RealmModule class *
-keep class io.realm.** { *; }
-dontwarn javax.**
-dontwarn io.realm.**
```

**Comment by nhachicha** [comment]
@zaki50 I don't think we need the above configuration, the user is using `1.0.0`

**Comment by zaki50** [comment]
@qsxiaogang Do you have default configuration of ProGuard in `app/build.gradle`?
I mean that `app/build.gradle` must contain `proguardFile getDefaultProguardFile('proguard-android.txt')`.

Could you check that?

**Comment by qsxiaogang** [comment]
@zaki50 yes ,here is my proguardfile configuration:
```
  buildTypes {
    release {
      signingConfig signingConfigs.config
      minifyEnabled true
      shrinkResources true
      proguardFile 'proguard-rules.pro'
    }
  }
```

**Comment by zaki50** [comment]
you need to use `getDefaultProguardFile('proguard-android.txt')` to include Android's default configuration.

Please try following configuration.

```
  buildTypes {
    release {
      signingConfig signingConfigs.config
      minifyEnabled true
      shrinkResources true
      proguardFile getDefaultProguardFile('proguard-android.txt') // <- added
      proguardFile 'proguard-rules.pro'
    }
  }
```

**Comment by Zhuinden** [comment]
@zaki50  you guys really need to document the fact that you need that line in your proguard config to make it work

**Comment by qsxiaogang** [comment]
@zaki50 i will try later.

**Comment by zaki50** [comment]
@Zhuinden Added a configuration for native methods to the ProGuard configuration file provided by Realm-java (#2963).

**Comment by stk1m1** [comment]
@qsxiaogang 

Has @zaki50's comment helped you on solving your issue?

**Comment by qsxiaogang** [comment]
@stk1m1 yes.
but Realm-java will be update in the next version with another way to solve ProGuard configuration .

**Comment by beeender** [comment]
Fix #2963 merged which will be released in `v1.0.1`
